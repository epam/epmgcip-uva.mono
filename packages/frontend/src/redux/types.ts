export const SET_USER_NAME = 'SET_USER_NAME';
export const SET_SHOW_MENU = 'SET_SHOW_MENU';