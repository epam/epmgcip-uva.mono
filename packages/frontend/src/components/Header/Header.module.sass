@use 'src/style/colors'
@import 'src/style/typography'

.headerWrapper 
    width: 100%
    max-width: 768px
    height: 46px
    background-color: colors.$white
    display: grid
    grid-template-columns: 55% 35% auto
    align-items: center
    justify-content: normal
    padding: 6px 12px
    margin: 0
    box-sizing: border-box
    border-bottom: 1px solid colors.$gray-light
    position: fixed
    top: 0
    z-index: 1

    .headerLogo
        height: 30px

    .userName
        @extend .H5
        color: colors.$violet
        text-align: end
