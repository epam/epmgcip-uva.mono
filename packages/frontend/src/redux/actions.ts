import { IEvent, IUser, ScrollDirection } from 'src/types';
import {
  ADD_USERS_TO_LIST,
  SET_SHOW_MENU,
  SET_EDITOR,
  UPDATE_USERS_LIST,
  SET_USER_SEARCH_INPUT,
  SET_MANAGE_USERS_SCROLL_SIZE,
  SET_MANAGE_USERS_SCROLL_DIRECTION,
  ADD_EVENTS_TO_LIST,
} from './types';

export const setEditor = (editor: IUser) => ({
  type: SET_EDITOR,
  payload: editor,
});

export const setMenu = (menuState: boolean) => ({
  type: SET_SHOW_MENU,
  payload: menuState,
});

export const addUsersToList = (users: IUser[]) => ({
  type: ADD_USERS_TO_LIST,
  payload: users,
});

export const updateUsersList = (users: IUser[]) => ({
  type: UPDATE_USERS_LIST,
  payload: users,
});

export const setManageUsersSearchInput = (searchInput: string) => ({
  type: SET_USER_SEARCH_INPUT,
  payload: searchInput,
});

export const setManageUsersScrollSize = (scrollSize: number) => ({
  type: SET_MANAGE_USERS_SCROLL_SIZE,
  payload: scrollSize,
})

export const setManageUsersScrollDirection = (scrollDirection: ScrollDirection) => ({
  type: SET_MANAGE_USERS_SCROLL_DIRECTION,
  payload: scrollDirection,
})

export const addEventsToList = (events: IEvent[]) => ({
  type: ADD_EVENTS_TO_LIST,
  payload: events,
});