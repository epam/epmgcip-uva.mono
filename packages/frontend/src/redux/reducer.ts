import { IAction, IState } from 'src/types';
import { SET_SHOW_MENU, SET_USER_NAME } from './types';
  
  
  const initialState: IState = {
    userName: '',
    isMenu: false,
    loading: false,
    error: null,
  };
  
  const rootReducer = (state = initialState, action: IAction) => {
    switch (action.type) {
      case SET_USER_NAME:
        return {
          ...state,
          userName: action.payload,
          error: null,
          loading: false,
        };
      case SET_SHOW_MENU:
        return {
            ...state,
            isMenu: action.payload,
          };
      default:
        return state;
    }
  };
  
  export default rootReducer;