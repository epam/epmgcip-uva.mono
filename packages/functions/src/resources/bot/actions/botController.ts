import { Telegraf } from 'telegraf';
import * as functions from 'firebase-functions/v1';

const bot = new Telegraf(functions.config().telegram.bot_token, {
  telegram: { webhookReply: true },
});

const registrationUrl = 'https://core.telegram.org/bots/webapps';

bot.command('register', async ctx => {
  await ctx.reply('Click the button below to register:', {
    reply_markup: {
      inline_keyboard: [
        [
          {
            text: 'Register',
            web_app: { url: registrationUrl },
          },
        ],
      ],
    },
  });
});

export default bot;
