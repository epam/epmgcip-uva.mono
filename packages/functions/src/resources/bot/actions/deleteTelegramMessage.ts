import axios from 'axios';
import * as admin from 'firebase-admin';
import { logger } from 'firebase-functions/v1';
import { TG_CHANNEL_ID } from '../../../constants/env.js';
import { TelegramBotCommands } from '../../../constants/telegram.js';
import { getHeaders, getTgCommandUrl } from '../../../utils/axios.js';

export const deleteTelegramMessage = async (eventData: admin.firestore.DocumentData) => {
  const telegramMessageId = eventData.telegramMessageId;
  try {
    await axios.post(
      getTgCommandUrl(TelegramBotCommands.deleteMessage),
      {
        chat_id: TG_CHANNEL_ID,
        message_id: telegramMessageId,
      },
      {
        headers: getHeaders(),
      },
    );
  } catch (err) {
    logger.error('Failed to delete message from Telegram', err);
  }
};
