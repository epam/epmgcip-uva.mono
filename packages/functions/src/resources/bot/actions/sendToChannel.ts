import axios from 'axios';
import admin from 'firebase-admin';
import { logger } from 'firebase-functions/v1';
import { EventStatus, FirebaseCollection } from 'uva-shared';
import { TG_CHANNEL_ID } from '../../../constants/env.js';
import { TelegramBotCommands } from '../../../constants/telegram.js';
import { getHeaders, getTgCommandUrl } from '../../../utils/axios.js';
import { createMessage } from '../dataModifyers/createMessage.js';
import { formatLanguageSpecificData } from '../dataModifyers/formatLanguageSpecificData.js';

export const sendToChannel = async (eventData: admin.firestore.DocumentData) => {
  const event = eventData;
  const formattedLanguageData = formatLanguageSpecificData(eventData.languageSpecificData);
  const message = createMessage(
    event,
    formattedLanguageData.description,
    formattedLanguageData.title,
    formattedLanguageData.eventPlace,
    formattedLanguageData.languageKeys,
  );
  if (event.status === EventStatus.Active && !event.telegramMessageId) {
    try {
      const response = await axios.post(
        getTgCommandUrl(TelegramBotCommands.sendPhoto),
        {
          chat_id: TG_CHANNEL_ID,
          caption: message,
          photo: eventData.imageUrl,
          parse_mode: 'HTML',
        },
        {
          headers: getHeaders(),
        },
      );

      await admin.firestore().collection(FirebaseCollection.Events).doc(event.id).update({
        telegramMessageId: response.data.result.message_id,
      });
    } catch (err) {
      logger.error('Failed to send a message to Telegram with event details.', err);
    }
  }
};
