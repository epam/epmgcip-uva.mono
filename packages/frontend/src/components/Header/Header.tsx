import css from './Header.module.sass';
import LogoSvg from 'src/assets/logo.svg';
import MenuSvg from 'src/assets/menu.svg';
import { Button } from 'src/components';
import { setMenu } from 'src/redux/actions';
import { useDispatch } from 'react-redux';
import { Dispatch } from '@reduxjs/toolkit';
import translation from 'src/translations/Russian.json';
import { ROOT_ROUTE } from 'src/constants';

interface HeaderProps {
  editorName?: string;
}

const maxNameLength = 25;

export const Header = ({ editorName }: HeaderProps) => {
  const dispatch: Dispatch = useDispatch();
  const currentRoute = window.location.pathname;
  const isMenuEnabled = currentRoute !== ROOT_ROUTE;
  const isLongName = editorName && editorName.length > maxNameLength;

  const handleShowMenu = () => {
    if (isMenuEnabled) {
      dispatch(setMenu(true));
    }
  };

  return (
    <div className={css.headerWrapper}>
      <img className={css.headerLogo} src={LogoSvg} />
      <div className={css.editorName}>
        {isLongName
          ? `${editorName?.slice(0, maxNameLength)}...`
          : editorName || translation.login}
      </div>
      <Button
        onClick={handleShowMenu}
        className={css.headerMenuButton}
        disabled={!isMenuEnabled}
      >
        <img className={css.headerLogo} src={MenuSvg} alt='Open Menu Button' />
      </Button>
    </div>
  );
};
