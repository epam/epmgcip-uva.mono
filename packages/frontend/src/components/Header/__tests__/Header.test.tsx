import { describe, expect, it, vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import { Header } from '../Header';
import { Provider } from 'react-redux';
import * as Actions from 'src/redux/actions';


const editorNameMock = 'Test User Name';
const observableMock = vi.fn();
const dispatchMock = <T,>() => vi.fn() as T;
const getStateMock = vi.fn();
const subscribeMock = vi.fn();
const replaceReducerMock = vi.fn();

const storeMock = {
  [Symbol.observable]: observableMock,
  dispatch: dispatchMock,
  getState: () => getStateMock,
  subscribe: () => subscribeMock,
  replaceReducer: () => replaceReducerMock,
};

vi.mock('react-redux', async () => {
  const actual = await vi.importActual('react-redux');

  return {
    ...actual,
    useDispach: () => ({ dispatch: dispatchMock }),
  };
});

describe('Testing: Header', () => {
  it('should render', () => {
    const { container } = render(
      <Provider store={storeMock}>
        <Header editorName={editorNameMock} />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should call dispatch after click', () => {
    const spyOnSetMenu = vi.spyOn(Actions, 'setMenu')

    render(
      <Provider store={storeMock}>
        <Header editorName={editorNameMock} />
      </Provider>
    );

    const renderedButton = screen.getByAltText('Open Menu Button');

    fireEvent.click(renderedButton);

    expect(spyOnSetMenu).toHaveBeenCalled();
  });
});
