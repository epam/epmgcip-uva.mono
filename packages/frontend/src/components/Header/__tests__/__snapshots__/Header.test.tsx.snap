// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing: Header > should render 1`] = `
<div>
  <div
    class="_headerWrapper_b11e5d"
  >
    <img
      class="_headerLogo_b11e5d"
      src="/packages/frontend/src/assets/logo.svg"
    />
    <div
      class="_editorName_b11e5d"
    >
      Test User Name
    </div>
    <button
      class="_button_6e8901"
    >
      <img
        alt="Open Menu Button"
        class="_headerLogo_b11e5d"
        src="/packages/frontend/src/assets/menu.svg"
      />
    </button>
  </div>
</div>
`;
